OpenRA.Mods.AS/Traits/TeleportNetworkPrimaryExit.cs

```diff
@@ -0,0 +1,107 @@
+﻿#region Copyright & License Information
+/*
+ * Modded by Boolbada of OP mod, from Primary production faciility designation logic.
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+/* Works without base engine modification */
+namespace OpenRA.Mods.AS.Traits
+{
+	static class TeleportNetworkPrimaryExitExts
+	{
+		public static bool IsPrimaryTeleportNetworkExit(this Actor a)
+		{
+			var pb = a.TraitOrDefault<TeleportNetworkPrimaryExit>();
+			return pb != null && pb.IsPrimary;
+		}
+	}
+
+	[Desc("Used with TeleportNetwork trait for primary exit designation")]
+	public class TeleportNetworkPrimaryExitInfo : ITraitInfo, Requires<TeleportNetworkInfo>
+	{
+		[GrantedConditionReference]
+		[Desc("The condition to grant to self while this is the primary building.")]
+		public readonly string PrimaryCondition = "primary";
+
+		[Desc("The speech notification to play when selecting a primary exit.")]
+		public readonly string SelectionNotification = "PrimaryBuildingSelected";
+
+		public object Create(ActorInitializer init) { return new TeleportNetworkPrimaryExit(init.Self, this); }
+	}
+
+	public class TeleportNetworkPrimaryExit : IIssueOrder, IResolveOrder, INotifyCreated
+	{
+		readonly TeleportNetworkPrimaryExitInfo info;
+		ConditionManager conditionManager;
+		int primaryToken = ConditionManager.InvalidConditionToken;
+		TeleportNetwork tni;
+
+		public bool IsPrimary { get; private set; }
+
+		public TeleportNetworkPrimaryExit(Actor self, TeleportNetworkPrimaryExitInfo info)
+		{
+			this.info = info;
+			IsPrimary = false;
+			tni = self.Trait<TeleportNetwork>();
```

Comment:
This will crash if there are multiple TN traits on the actor.

The TN handling of exits already imply TN traits should use the same exits (which I'm fine with) but the handling of TNs within this trait (and the ext above) isn't prepped for multiple TN traits case.